### DISCUSSION TITLE
onchange function

### DISCUSSION THREAD

**Discussion by PaulMcF1987** [Budibase/budibase#6930] [Help]
Not sure if this is a bug or if I am doing something wrong.

Again, for the calculator that I have mentioned a few times...

I have set up all fields with an `onchange` function so that when I edit a field, it then calculates a value...

Because the value can be determined by a combination of different fields, each field has multiple onchange formulas so that it recalculates the other fields as well... 

The problem I am finding is...

`Box A + Box B + Box C generates a service value in box Z` depending on the inputs to A, B & C

`Box D then generates a price in Box Y`

In Box A, B & C the First Formula Generates the output to Box Z...

The second formula is then intended to use the output from Box Z to then calculate Box Y based on the input in Box D.

However, This does not work the first time you select the input in on of the boxes, It only runs the first function. I then need to select the input again in order for it to run the second function.

I was hoping the fact that the functions could be ordered would mean that having a setup where by

    Function 1 populates box Z
    Function 2 uses the entry form Box Z

Would be fine because Box Z should be populated by the time function 2 runs.
Could someone shed any light on this?

--------------------------------------------------------------------------------------------------------------------------------------

I tried adding an onchange function to Box Z so that it automatically updates Box Y when changes are made but it does not work.

It appears that if a box is updated as a result of a funtion, it does not register the onchange function

**Comment by mmalessa80** [comment]
Maybe a couple of things to observe here, on change in the case of budibase appears to only trigger with a user change in a field, not a programmatic change. So your second to last paragraph would suggest you have onchange events for A, B, and C to recalculate Z, and then you have an onchange event on Z to recalculate Y. So user entry in A, B, or C will calculate Z, but then Z doesn’t ever get physical user input so that onchange would not fire and therefore Y does not get updated. I would suggest trying to incorporate the calculation for Z and Y in the onchange event of A, B, and C. Also, in your calculations make sure to use ‘field value’ instead of ‘data provider field’ as the first will be the value the field was updated to whereas the second is the last saved value for that column.

**Reply by PaulMcF1987** [reply]
@mmalessa80 I did initially have 2 functions in the onchange of A, B and C. However, it would only run one function, then when I selected the option again, it would run the second function.

I couldnt incorporate both into the same function as they are required to update 2 different boxes.

I think the problem occurs because when there are 2 functions in the on onchange function. The second function relies on the input from function 1.
If both functions run at the same time, then it will either be running based on an empty entry or the old entry before function 1 has populated the new entry

if that makes sense?

Im not sure the best way to overcome this.

**Reply by mmalessa80** [reply]
Can you show screenshots of the dual function setup?
